mmocr/core/evaluation/utils.py

```diff
@@ -133,53 +133,85 @@ def points2polygon(points):
     assert (points.size % 2 == 0) and (points.size >= 8)
 
     point_mat = points.reshape([-1, 2])
-    return plg.Polygon(point_mat)
+    return plg(point_mat)
 
 
-def poly_intersection(poly_det, poly_gt):
+def poly_intersection(poly_det, poly_gt, invalid_ret=0, return_poly=False):
     """Calculate the intersection area between two polygon.
 
     Args:
         poly_det (Polygon): A polygon predicted by detector.
         poly_gt (Polygon): A gt polygon.
-
+        invalid_ret (int|float): The return value when invalid polygon exists.
+        return_poly (bool): Whether to return the polygon of the intersection
+                            area.
     Returns:
         intersection_area (float): The intersection area between two polygons.
+        poly_obj (Polygon, optional): The Polygon object of the intersection
+                                    area. Set as `None` if the input is
+                                    invalid.
     """
-    assert isinstance(poly_det, plg.Polygon)
-    assert isinstance(poly_gt, plg.Polygon)
+    assert isinstance(poly_det, plg)
+    assert isinstance(poly_gt, plg)
 
-    poly_inter = poly_det & poly_gt
-    if len(poly_inter) == 0:
-        return 0, poly_inter
-    return poly_inter.area(), poly_inter
+    if poly_det.is_valid and poly_gt.is_valid:
+        poly_obj = poly_det.intersection(poly_gt)
+        if return_poly:
+            return poly_obj.area, poly_obj
+        else:
+            return poly_obj.area
```

Comment:
can it return negative area?

Reply:
No, it always returns a non-negative number. If an invalid polygon is given, shapely will raise an exception.